pkg/metrics/metrics_test.go

```diff
@@ -0,0 +1,217 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"io/ioutil"
+	"math"
+	"net/http"
+	"os"
+	"reflect"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	cmg "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var (
+	mgr MetricsManager
+	srv *http.Server
+	wg  *sync.WaitGroup
+)
+
+const (
+	httpPattern = "/metrics"
+	addr        = ":12134"
+	httpAddr    = "http://localhost" + addr + httpPattern
```

Comment:
This failed on my setup too:

```
--- FAIL: TestStartServing (0.00s)
    metrics_test.go:77: failed to get response from server Get http://localhost:12134/metrics: dial tcp 127.0.0.1:12134: connect: connection refused
```

Reply:
Hmm, seems very flaky.  I ran the test a few times and finally it passed.

Reply:
yes, its failing constantly, I will need to fix it.
could you share with me your set up? do you have a full travis setup?
I ran:
make -k all test
it worked fine.

Reply:
This is just on my dev machine.  I go to the metrics directory and ran "go test".

Reply:
Tried "make -k all test" but it gave me the same error.

Reply:
this is now fixed. @oleksiys helped figured it out. The root cause was the call "ListenAndServe" does two major things: Listen on the provided address, and start serving, and it's done in a separate goroutine. When the test code tries to call "Get", the server might not necessarily have been initialized yet and causes connection refused error.